source/common/http/websocket/ws_handler_impl.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/http/filter.h"
+#include "envoy/network/connection.h"
+#include "envoy/network/filter.h"
+#include "envoy/upstream/cluster_manager.h"
+#include "envoy/upstream/upstream.h"
+
+#include "common/common/logger.h"
+#include "common/network/filter_impl.h"
+
+namespace Envoy {
+namespace Http {
+namespace WebSocket {
+
+/**
+ * An implementation of a WebSocket proxy based on TCP proxy. This filter will instantiate a
+ * new outgoing TCP connection using the defined load balancing proxy for the configured cluster.
+ * All data will be proxied back and forth between the two connections, without any knowledge of
+ * the underlying WebSocket protocol.
+ *
+ * N.B. This class implements Network::ReadFilter interfaces purely for sake of consistency
+ * with TcpProxy filter. WsHandlerImpl it is not used as a network filter in any way.
+ */
+class WsHandlerImpl : public Network::ReadFilter, Logger::Loggable<Logger::Id::websocket> {
```

Comment:
Very quick comment: I would like to see if there is a way to share the code with tcp_proxy, either via a shared utility class or consuming tcp_proxy internally much like async_client consumes the router filter.